onnx/cpp2py_export.cc

```diff
@@ -115,13 +115,215 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
   py::register_exception<SchemaError>(defs, "SchemaError");
 
   py::class_<OpSchema> op_schema(defs, "OpSchema", "Schema of an operator.");
-  op_schema.def_property_readonly("file", &OpSchema::file)
+
+  // Define the class enums first because they are used as default values in function definitions
+  py::enum_<OpSchema::FormalParameterOption>(op_schema, "FormalParameterOption")
+      .value("Single", OpSchema::Single)
+      .value("Optional", OpSchema::Optional)
+      .value("Variadic", OpSchema::Variadic);
+
+  py::enum_<OpSchema::DifferentiationCategory>(op_schema, "DifferentiationCategory")
+      .value("Unknown", OpSchema::Unknown)
+      .value("Differentiable", OpSchema::Differentiable)
+      .value("NonDifferentiable", OpSchema::NonDifferentiable);
+
+  py::enum_<AttributeProto::AttributeType>(op_schema, "AttrType")
+      .value("FLOAT", AttributeProto::FLOAT)
+      .value("INT", AttributeProto::INT)
+      .value("STRING", AttributeProto::STRING)
+      .value("TENSOR", AttributeProto::TENSOR)
+      .value("GRAPH", AttributeProto::GRAPH)
+      .value("FLOATS", AttributeProto::FLOATS)
+      .value("INTS", AttributeProto::INTS)
+      .value("STRINGS", AttributeProto::STRINGS)
+      .value("TENSORS", AttributeProto::TENSORS)
+      .value("GRAPHS", AttributeProto::GRAPHS)
+      .value("SPARSE_TENSOR", AttributeProto::SPARSE_TENSOR)
+      .value("SPARSE_TENSORS", AttributeProto::SPARSE_TENSORS)
+      .value("TYPE_PROTO", AttributeProto::TYPE_PROTO)
+      .value("TYPE_PROTOS", AttributeProto::TYPE_PROTOS);
+
+  py::enum_<OpSchema::SupportType>(op_schema, "SupportType")
+      .value("COMMON", OpSchema::SupportType::COMMON)
+      .value("EXPERIMENTAL", OpSchema::SupportType::EXPERIMENTAL);
+
+  py::class_<OpSchema::Attribute>(op_schema, "Attribute")
+      .def(
+          py::init([](std::string name, AttributeProto::AttributeType type, std::string description, bool required) {
+            // Construct an attribute.
+            // Use a lambda to swap the order of the arguments to match the Python API
+            return OpSchema::Attribute(name, description, type, required);
+          }),
+          py::arg("name"),
+          py::arg("type"),
+          py::arg("description") = "",
+          py::kw_only(),
+          py::arg("required") = true)
+      .def(
+          py::init([](std::string name, const py::object& default_value, std::string description) {
+            // Construct an attribute with a default value.
+            // Attributes with default values are not required
+            auto bytes = default_value.attr("SerializeToString")().cast<py::bytes>();
+            AttributeProto proto{};
+            ParseProtoFromPyBytes(&proto, bytes);
+            return OpSchema::Attribute(name, description, proto);
```

Comment:
I remember reading that the compiler will know to move automatically in some cases. Is there a good rule of thumb on when that will happen?

Reply:
we can start with this advice from [here](https://herbsutter.com/2020/02/17/move-simply/):

>To pass a named object `a` as an argument to a `&&` “move” parameter (rvalue reference parameter), write `std::move(a)`. That’s pretty much the only time you should write std::move, because C++ already uses move automatically when copying from an object it knows will never be used again, such as a temporary object or a local variable being returned or thrown from a function.

a named object is an lvalue, so we would need to cast it to an rvalue reference which is just what std::move does.

in this code, we also deal with functions which take a string (or some other type we might not want to copy) by value, for example, `OpSchema::Attribute::Attribute()`. this way, we can avoid writing two versions of this constructor, one taking a `const string&` (where we copy the parameter) and the other taking a `string&&` (where we move from the parameter). so we can use `std::move(variable_name)` here, in the spirit of the advice above, to move into the by-value parameter instead of copy into it.

some more details on this pass-by-value pattern [here](https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference).

as another example, the compiler will move without us writing std::move if we pass an rvalue as the parameter:
`Attribute attr(GetName(), ...);`